### PR TITLE
jruby: 9.3.9.0 -> 9.4.3.0

### PR DESCRIPTION
**Pull request by marsam**
###### Description of changes

>JRuby 9.4.x targets Ruby 3.1 compatibility.

https://github.com/jruby/jruby/releases/tag/9.4.3.0
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes (or backporting 23.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/interpreters/jruby/default.nix [hunk 1]
```diff
@@ -1,27 +1,28 @@
-{ lib, stdenv, callPackage, fetchurl, makeWrapper, jre }:
+{ lib, stdenv, callPackage, fetchurl, makeBinaryWrapper, jre }:
 
 let
 # The version number here is whatever is reported by the RUBY_VERSION string
-rubyVersion = callPackage ../ruby/ruby-version.nix {} "2" "5" "7" "";
+rubyVersion = callPackage ../ruby/ruby-version.nix {} "3" "1" "4" "";
 jruby = stdenv.mkDerivation rec {
```
Comment (drupol) [inline pkgs/development/interpreters/jruby/default.nix:6]: Hi!

I added a non-blocking comment, feel free to implement it if you want, if not, just let me know and I'll merge this immediately.

**Review by drupol — COMMENTED** [review]
Hi!

I added a non blocking comment. Just let me know if you're willing to tackle it in this PR as well. If not, I'll merge this immediately.

I believe that you could even refactor the `jruby.overrideAttrs` part.

**Comment by fzakaria** [comment]
Did the docs and bash output change + removed ?
Other changes look good but haven't tested.

File: pkgs/development/interpreters/jruby/default.nix (hunk 1 above)
Comment (fzakaria) [reply pkgs/development/interpreters/jruby/default.nix:6]: What's the comment ?

Comment (drupol) [reply pkgs/development/interpreters/jruby/default.nix:6]: Looks like it was lost, I don't know why. Let me put it back.

File: pkgs/development/ruby-modules/gem/default.nix [hunk 2]
```diff
@@ -136,7 +136,7 @@ stdenv.mkDerivation ((builtins.removeAttrs attrs ["source"]) // {
 
   # As of ruby 3.0, ruby headers require -fdeclspec when building with clang
   # Introduced in https://github.com/ruby/ruby/commit/0958e19ffb047781fe1506760c7cbd8d7fe74e57
-  env.NIX_CFLAGS_COMPILE = toString (lib.optionals (stdenv.cc.isClang && lib.versionAtLeast ruby.version.major "3") [
+  env.NIX_CFLAGS_COMPILE = toString (lib.optionals (ruby.rubyEngine == "ruby" && stdenv.cc.isClang && lib.versionAtLeast ruby.version.major "3") [
```
Comment (fzakaria) [inline pkgs/development/ruby-modules/gem/default.nix:139]: is this even needed?

**Review by fzakaria — COMMENTED** [review]
looks sensible -- can you include some output that it ran succesfully?

**Comment by marsam** [comment]
>Did the docs and bash output change + removed ?

I moved `$out/docs` to `$out/share/jruby/docs` which seems more common in nixpkgs. Let me know if you prefer the former, I don't mind reverting it.
Also, upstream removed jruby.bash.

File: pkgs/development/interpreters/jruby/default.nix (hunk 1 above)
Comment (marsam) [reply pkgs/development/interpreters/jruby/default.nix]: Thanks for reviewing. I refactored the expression to use the `finalAttrs` pattern.

**Review by fzakaria — APPROVED** [review]

**Review by drupol — APPROVED** [review]

**Comment by drupol** [comment]
Thanks for fixing everything!

**Pull request merged**